lib/src/value/color.ts

```diff
@@ -3,371 +3,1182 @@
+    case 'srgb':
+    case 'srgb-linear':
+    case 'display-p3':
+    case 'a98-rgb':
+    case 'prophoto-rgb':
+    case 'rec2020':
+      valid = ['red', 'green', 'blue'].includes(channel);
+      break;
+    case 'hsl':
+      valid = ['hue', 'saturation', 'lightness'].includes(channel);
+      break;
+    case 'hwb':
+      valid = ['hue', 'whiteness', 'blackness'].includes(channel);
+      break;
+    case 'lab':
+    case 'oklab':
+      valid = ['lightness', 'a', 'b'].includes(channel);
+      break;
+    case 'lch':
+    case 'oklch':
+      valid = ['lightness', 'chroma', 'hue'].includes(channel);
+      break;
+    case 'xyz':
+    case 'xyz-d65':
+    case 'xyz-d50':
+      valid = ['x', 'y', 'z'].includes(channel);
+      break;
+  }
+  if (!valid) {
+    throw valueError(
+      `Unknown channel name "${channel}" for color space "${space}".`
+    );
+  }
+}
+
+/** Determine whether the given space is a polar color space. */
+function isPolarColorSpace(space: KnownColorSpace): space is PolarColorSpace {
+  switch (space) {
+    case 'hsl':
+    case 'hwb':
+    case 'lch':
+    case 'oklch':
+      return true;
+    default:
+      return false;
+  }
+}
+
+/**
+ * Convert from ColorJS coordinates (which use `NaN` for missing components, and
+ * a range of `0-1` for `rgb` channel values) to Sass Color coordinates (which
+ * use `null` for missing components, and a range of `0-255` for `rgb` channel
+ * values).
+ */
+function decodeCoordsFromColorJs(
+  coords: [number, number, number], // ColorJS coordinates
+  isRgb = false // Whether this color is in the `rgb` color space
+): [number | null, number | null, number | null] {
+  let newCoords = coords;
+  // If this color is in the `rgb` space, convert channel values to `0-255`
+  if (isRgb) newCoords = newCoords.map(coordToRgb) as [number, number, number];
+  // Convert `NaN` values to `null`
+  return newCoords.map(NaNtoNull) as [
+    number | null,
+    number | null,
+    number | null,
+  ];
 }
 
-interface HslColor {
-  hue: number;
-  saturation: number;
-  lightness: number;
-  alpha?: number;
+/** Returns `true` if `val` is a `number` or `null`. */
+function isNumberOrNull(val: undefined | null | number): val is number | null {
+  return val === null || typeof val === 'number';
 }
 
-interface HwbColor {
-  hue: number;
-  whiteness: number;
-  blackness: number;
-  alpha?: number;
+/**
+ * Emit deprecation warnings when legacy color spaces set `alpha` or channel
+ * values to `null` without explicitly setting the `space`.
+ */
+function checkChangeDeprecations(
+  options: {
+    [key in ChannelName]?: number | null;
+  },
+  channels: ChannelName[]
+): void {
+  if (options.alpha === null) emitNullAlphaDeprecation();
+  for (const channel of channels) {
+    if (options[channel] === null) emitColor4ApiChangeNullDeprecation(channel);
+  }
+}
+
+/** Warn users about legacy color channel getters. */
+function emitColor4ApiGetterDeprecation(name: string): void {
```

Comment:
All of the color-4-api deprecation warnings need some sort of handling for `silenceDeprecations` and `fatalDeprecations` since the host can't rely on the compiler to do the processing for any warnings emitted directly by the host.

Reply:
I think it might be worth just merging this anyway so that we can get rid of the feature branch hassle, and then updating this along with your incoming change.